leader/src/cli.rs

```diff
@@ -18,6 +23,63 @@ pub(crate) struct Cli {
     pub(crate) prover_state_config: CliProverStateConfig,
 }
 
+#[derive(Clone, Debug)]
+pub(super) enum BlockNumbers {
+    Single(u64),
+    RangeInclusive(RangeInclusive<u64>),
+}
+
+impl FromStr for BlockNumbers {
+    type Err = anyhow::Error;
+
+    /// Parses input block numbers to determine a specific block or a range of
+    /// blocks.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::from_str_intern(s)
+            .with_context(|| {
+                format!(
+                    "Expected a single value or a range, but instead got \"{}\".",
+                    s
+                )
+            })
+            .map_err(|e| anyhow!("{e:#}"))
+    }
+}
+
+impl BlockNumbers {
+    fn from_str_intern(s: &str) -> anyhow::Result<Self> {
+        // Did we get passed a single value?
+        if let Ok(v) = s.parse::<u64>() {
+            return Ok(Self::Single(v));
+        }
+
+        // Check if it's a range.
+        let mut range_vals = s.split("..=");
+
+        let start = Self::next_and_try_parse(&mut range_vals)?;
+        let end = Self::next_and_try_parse(&mut range_vals)?;
+
+        if range_vals.count() > 0 {
+            return Err(anyhow!(
+                "Parsed a range but there were unexpected characters afterwards!"
+            ));
+        }
+
+        Ok(Self::RangeInclusive(start..=end))
+    }
+
+    fn next_and_try_parse(range_vals: &mut Split<&str>) -> anyhow::Result<u64> {
+        let unparsed_val = range_vals
+            .next()
+            .with_context(|| "Parsing a value as a `RangeInclusive`")?;
+        let res = unparsed_val
+            .parse()
+            .with_context(|| format!("Parsing the range val \"{unparsed_val}\" into a usize"))?;
```

Comment:
nit: I don't think we ever used this format of `... {foo}..` in formatting in other places of the codebase (or even in `zk_evm`). Perhaps out of consistency we should go with this?
```suggestion
            .with_context(|| format!("Parsing the range val \"{}\" into a usize", unparsed_val))?;
```